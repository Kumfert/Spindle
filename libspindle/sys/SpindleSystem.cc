//
// SpindleSystem.cc
//
// $Id: SpindleSystem.cc,v 1.2 2000/02/18 01:32:01 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify , distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//

#include "spindle/SpindleSystem.h"

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

static int count = 0;

static FILE* TraceFile = stderr;
static FILE* WarnFile = stderr;
static FILE* ErrorFile = stderr;

static unsigned int TraceLevel = 0;
static PersistanceRegistry * Registry = 0;
static StackTrace * Stack = 0;

static char* currentFunction;
static char* currentFile;
static int currentLine;

PersistanceRegistry& 
SpindleSystem::registry() {
  return *Registry;
}

StackTrace& 
SpindleSystem::stack() {
  return *Stack;
}

void
SpindleSystem::startSystem() {
  if (count++ == 0) {
    Registry = new PersistanceRegistry();
    Stack = new StackTrace();
  }
}

void
SpindleSystem::stopSystem() {
  if ( --count==0 ) {
    delete Registry;
    delete Stack;
  }
}



void
SpindleSystem::setWhere( char* func, char* file, int line ) {
  currentFunction = func;
  currentFile = file;
  currentLine = line;
}

void 
SpindleSystem::setTraceFile( FILE* traceFile ) {
  TraceFile = traceFile;
}

void 
SpindleSystem::setWarnFile( FILE* warnFile ) {
  WarnFile = warnFile;
}


void 
SpindleSystem::setErrorFile( FILE* errorFile ) {
  ErrorFile = errorFile;
}

void 
SpindleSystem::setTraceLevel( unsigned int traceLevel ) {
  if ( traceLevel >= 0 ) {
    TraceLevel = traceLevel;
  }
}

unsigned int 
SpindleSystem::getTraceLevel() { 
  return TraceLevel;
}

//
// Now implement the print functions for standard libraries.
//
//
#if SPINDLE_TARGET == std

void 
SpindleSystem::assert_that( bool arg, const char* format, ... ) {
  if ( (! arg)  && (ErrorFile != 0 ) ) {
    va_list ap;
    va_start( ap, format);
    fprintf(ErrorFile, "\nASSERT FAILED:  ");
    vfprintf(ErrorFile, format, ap);
#ifdef SPINDLE_ENABLE_STACK_TRACE
    SpindleSystem::stack().dump( ErrorFile );
    SpindleSystem::stack().reset();
#else
    SpindleSystem::printWhere( ErrorFile );
#endif
    va_end(ap);
    // exit(-1);
  }
}

void 
SpindleSystem::error( unsigned long errorCode, const char* format, ... ) {
  va_list ap;
  va_start( ap, format );
  if ( ErrorFile != 0 ) {
    fprintf(ErrorFile, "\nERROR(%ld):  ",errorCode);
    vfprintf(ErrorFile, format, ap);
#ifdef SPINDLE_ENABLE_STACK_TRACE
    SpindleSystem::stack().dump( ErrorFile );
#else
    printWhere( ErrorFile );
#endif
  }
  va_end(ap);
}

merr
SpindleSystem::mError(  int line, char* file, char* func, merr errcode, char* msg ) {
  if ( ErrorFile != 0 ) {
    fprintf(ErrorFile, "\nERROR(%d): %s in \"%s\" %s[%d]",errcode, msg, func, file, line );
  }
  return errcode;
}

void 
SpindleSystem::warn_if( bool arg, const char* format, ... ) {
  if ( arg && ( WarnFile != 0 ) ) {
    va_list ap;
    va_start( ap, format);
    fprintf(WarnFile, "\nWARNING:  ");
    vfprintf(WarnFile, format, ap);
#ifdef SPINDLE_ENABLE_STACK_TRACE
    SpindleSystem::stack().dump( WarnFile );
#else 
    SpindleSystem::printWhere( WarnFile );
#endif
    va_end(ap);
  }
}

void 
SpindleSystem::warning( const char* format, ... ) {
  va_list ap;
  va_start( ap, format);
  if ( WarnFile != 0 ) {
    fprintf(WarnFile, "\nWARNING:  ");
    vfprintf(WarnFile, format, ap);
#ifdef SPINDLE_ENABLE_STACK_TRACE
    SpindleSystem::stack().dump( WarnFile );
#else
    printWhere( ErrorFile );
#endif
  }
  va_end(ap);
}

void 
SpindleSystem::trace( unsigned int traceLevel, const char* format, ... ) {
  if ( traceLevel & TraceLevel ) {
    va_list ap;
    va_start( ap, format);
    //    const StackInfo& si = SpindleSystem::stack().top();
    //    if ( traceLevel != SPINDLE_TRACE_STACK ) {
    //      fprintf(TraceFile, "\nTrace(%d) %s [%d] - \"%s\"\n\t",traceLevel, si.file, si.line, si.function );
    //    }
    if ( TraceFile != 0 ) {
      vfprintf(TraceFile, format, ap);
    }
    va_end(ap);
  }
}

void 
SpindleSystem::printf( const char* format, ... ) {
  va_list ap;
  va_start( ap, format);
  if ( TraceFile != 0 ) {
    vfprintf(TraceFile, format, ap);
  }
  va_end(ap);
}

void 
SpindleSystem::printWhere( FILE * fp ) {
  if ( (currentFunction != 0) && (currentFile != 0 ) && (fp != 0 ) ) {
    fprintf(fp, "\n \"%s\" in %s-%d\n",currentFunction, currentFile, currentLine );
  }
}

//
// else make a library using matlab constructs...
//
//
#elif SPINDLE_TARGET == matlab 

extern "C" {
#include "mex.h"
}
static char buffer[512*8];

void 
SpindleSystem::assert_that( bool arg, const char* format, ... ) {
  if ( ! arg ) {
    va_list ap;
    va_start( ap, format);
    if ( ErrorFile == 0 )  { // print via Matlab's messaging system.
      int i = sprintf( buffer, "\nASSERT FAILED:  ");
      vsprintf( buffer+i, format, ap );
      mexWarnMsgTxt( buffer );
    } else {  // assume that a file is open
      fprintf(ErrorFile, "\nASSERT FAILED:  ");
      vfprintf(ErrorFile, format, ap);
#ifdef SPINDLE_ENABLE_STACK_TRACE
      SpindleSystem::stack().dump( ErrorFile );
      SpindleSystem::stack().reset();
#else
      SpindleSystem::printWhere( ErrorFile );
#endif
    }
    va_end(ap);
    // exit(-1);
  }
}

void 
SpindleSystem::error( unsigned long errorCode, const char* format, ... ) {
  va_list ap;
  va_start( ap, format );
  if ( ErrorFile == 0 ) {
      int i = sprintf( buffer, "\nASSERT FAILED:  ");
      vsprintf( buffer+i, format, ap );
      mexWarnMsgTxt( buffer );
  } else if ( ErrorFile != 0 ) {
    fprintf(ErrorFile, "\nERROR(%ld):  ",errorCode);
    vfprintf(ErrorFile, format, ap);
#ifdef SPINDLE_ENABLE_STACK_TRACE
    SpindleSystem::stack().dump( ErrorFile );
#else
    printWhere( ErrorFile );
#endif
  }
  va_end(ap);
}

merr
SpindleSystem::mError(  int line, char* file, char* func, merr errcode, char* msg ) {
  if ( ErrorFile == 0 ) {
      vsprintf( buffer, "\nERROR(%d): %s in \"%s\" %s[%d]",errcode, msg, func, file, line );
      mexWarnMsgTxt( buffer );
  } else if ( ErrorFile != 0 ) {
    fprintf(ErrorFile, "\nERROR(%d): %s in \"%s\" %s[%d]",errcode, msg, func, file, line );
  }
  return errcode;
}

void 
SpindleSystem::warn_if( bool arg, const char* format, ... ) {
  if ( arg ) {
    va_list ap;
    va_start( ap, format);
    if ( WarnFile == 0 ) {
      int i = sprintf( buffer, "\nWARNING:   ");
      vsprintf( buffer+i, format, ap );
      mexWarnMsgTxt( buffer );
    } else {
      fprintf(WarnFile, "\nWARNING:  ");
      vfprintf(WarnFile, format, ap);
#ifdef SPINDLE_ENABLE_STACK_TRACE
      SpindleSystem::stack().dump( WarnFile );
#else 
      SpindleSystem::printWhere( WarnFile );
#endif
    }
    va_end(ap);
  }
}

void 
SpindleSystem::warning( const char* format, ... ) {
  va_list ap;
  va_start( ap, format);
  if ( WarnFile == 0 ) {
    int i = sprintf( buffer, "\nWARNING:  ");
    vsprintf( buffer+i, format, ap );
    mexWarnMsgTxt( buffer );
  } else if ( WarnFile != 0 ) {
    fprintf(WarnFile, "\nWARNING:  ");
    vfprintf(WarnFile, format, ap);
#ifdef SPINDLE_ENABLE_STACK_TRACE
    SpindleSystem::stack().dump( WarnFile );
#else
    printWhere( ErrorFile );
#endif
  }
  va_end(ap);
}

void 
SpindleSystem::trace( unsigned int traceLevel, const char* format, ... ) {
  if ( traceLevel & TraceLevel ) {
    va_list ap;
    va_start( ap, format);
    //    const StackInfo& si = SpindleSystem::stack().top();
    //    if ( traceLevel != SPINDLE_TRACE_STACK ) {
    //      fprintf(TraceFile, "\nTrace(%d) %s [%d] - \"%s\"\n\t",traceLevel, si.file, si.line, si.function );
    //    }
    if ( TraceFile == 0 ) {
      vsprintf( buffer, format, ap );
      mexPrintf( "%s", buffer );
    } else if ( TraceFile != 0 ) {
      vfprintf(TraceFile, format, ap);
    }
    va_end(ap);
  }
}

void 
SpindleSystem::printf( const char* format, ... ) {
  va_list ap;
  va_start( ap, format);
  if ( TraceFile == 0 ) {
    vsprintf( buffer, format, ap );
    mexPrintf( "%s", buffer );
  } else  if ( TraceFile != 0 ) {
    vfprintf(TraceFile, format, ap);
  }
  va_end(ap);
}

void 
SpindleSystem::printWhere( FILE * fp ) {
  if ( (currentFunction != 0) && (currentFile != 0 ) && (fp != 0 ) ) {
    fprintf(fp, "\n \"%s\" in %s-%d\n",currentFunction, currentFile, currentLine );
  }
  if ( (currentFunction != 0) && (currentFile != 0 ) && (fp == 0 ) ) {
    sprintf( buffer, "\n \"%s\" in %s-%d\n",currentFunction, currentFile, currentLine );
    mexPrintf( "%s", buffer );
  }
}

#elif  SPINDLE_TARGET == petsc 

extern "C" {
#include "petsc.h"
}
static char buffer[512*8];

void 
SpindleSystem::assert_that( bool arg, const char* format, ... ) {
  if ( ! arg  && ( ErrorFile != 0 ) ) {
    va_list ap;
    va_start( ap, format);
    int i = sprintf( buffer, "\nASSERT FAILED:  ");
    vsprintf( buffer+i, format, ap);
    PetscFPrintf( PETSC_COMM_SELF, ErrorFile, "%s", buffer );
    SpindleSystem::printWhere( ErrorFile );
    va_end(ap);
    // exit(-1);
  }
}

void 
SpindleSystem::error( unsigned long errorCode, const char* format, ... ) {
  va_list ap;
  va_start( ap, format );
  if ( ErrorFile != 0 ) {
    int i = sprintf( buffer, "\nERROR(%ld):  ", errorCode );
    vsprintf( buffer+i, format, ap);
    PetscFPrintf( PETSC_COMM_SELF, ErrorFile, "%s", buffer );
    SpindleSystem::printWhere( ErrorFile );
    printWhere( ErrorFile );
  }
  va_end(ap);
}

merr
SpindleSystem::mError(  int line, char* file, char* func, merr errcode, char* msg ) {
  if ( ErrorFile != 0 ) {
    PetscFPrintf( PETSC_COMM_SELF, ErrorFile, "\nERROR(%d): %s in \"%s\" %s[%d]",errcode, msg, func, file, line );
  }
  return errcode;
}

void 
SpindleSystem::warn_if( bool arg, const char* format, ... ) {
  if ( arg ) {
    va_list ap;
    va_start( ap, format);
    if ( WarnFile != 0 ) {
      int i = sprintf( buffer, "\nWARNING:  ");
      vsprintf( buffer+i, format, ap );
      PetscFPrintf( PETSC_COMM_SELF, WarnFile, "%s", buffer );
      SpindleSystem::printWhere( WarnFile );
    }
    va_end(ap);
  }
}

void 
SpindleSystem::warning( const char* format, ... ) {
  va_list ap;
  va_start( ap, format);
  if ( WarnFile != 0 ) {
    int i = sprintf( buffer, "\nWARNING:  ");
    vsprintf(buffer+i, WarnFile, format, ap);
    PetscFPrintf( PETSC_COMM_SELF, WarnFile, "%s", buffer );
    printWhere( WarnFile );
  }
  va_end(ap);
}

void 
SpindleSystem::trace( unsigned int traceLevel, const char* format, ... ) {
  if ( traceLevel & TraceLevel ) {
    va_list ap;
    va_start( ap, format);
    //    const StackInfo& si = SpindleSystem::stack().top();
    //    if ( traceLevel != SPINDLE_TRACE_STACK ) {
    //      fprintf(TraceFile, "\nTrace(%d) %s [%d] - \"%s\"\n\t",traceLevel, si.file, si.line, si.function );
    //    }
    if ( TraceFile != 0 ) {
      vsprintf(buffer , format, ap);
      PetscFPrintf( PETSC_COMM_SELF, TraceFile, "%s", buffer );
    }
    va_end(ap);
  }
}

void 
SpindleSystem::printf( const char* format, ... ) {
  va_list ap;
  va_start( ap, format);
  vsprintf( buffer , format, ap );
  PetscPrintf( PETSC_COMM_SELF,  "%s", buffer );
  va_end(ap);
}

void 
SpindleSystem::printWhere( FILE * fp ) {
  if ( (currentFunction != 0) && (currentFile != 0 ) && (fp != 0 ) ) {
    PetscFPrintf( PETSC_COMM_SELF, fp, "\n \"%s\" in %s-%d\n",currentFunction, currentFile, currentLine );
  }
  if ( (currentFunction != 0) && (currentFile != 0 ) && (fp == 0 ) ) {
    PetscPrintf( PETSC_COMM_SELF, "\n \"%s\" in %s-%d\n",currentFunction, currentFile, currentLine );
  }
}

#else
#error Error do not recognize SPINDLE TARGET of SPINDLE_TARGET
#endif
