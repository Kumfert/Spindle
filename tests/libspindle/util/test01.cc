//
// src/util/drivers/test/test01.cc 
//
// $Id: test01.cc,v 1.2 2000/02/18 01:32:11 kumfert Exp $
// 
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1998, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
// This exercises some of the BreadthFirstSearch capabilities
//

#include <iostream.h>
#include "spindle/SpindleSystem.h"
#include "spindle/BreadthFirstSearch.h"
#include "samples/small_mesh.h"

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

static void printAll(const BreadthFirstSearch& bfs);

#ifdef __FUNC__
#undef __FUNC__
#endif

int main() {
  Graph graph(nvtxs,Aptr,Aind);
  graph.validate();
  if ( ! graph.isValid() ) {
    cerr << "Cannot validate graph" << endl;
  }

  cout << "Testing the adjacency iterators over the graph" << endl;
  { for(int i=0; i<graph.size(); i++) {
    cout << "adj(" << i << ") :";
    for(Graph::const_iterator cur=graph.begin_adj(i),stop=graph.end_adj(i); 
	cur != stop; ++cur) {
      cout << *cur << ", ";
    }
    cout << endl;
  }}
  
  cout << endl << "Creating the Breadth First Search (BFS) object ... ";
  BreadthFirstSearch bfs(&graph);
  cout << " done. " << endl;
  printAll( bfs );
  
  cout << "Doing BFS from vertex 2 ... ";
  bfs.setRoot(2);
  if ( bfs.execute() ) {
    cout << " done. " << endl;
  } else {
    cout << " problem detected ! " << endl;
  }
  printAll( bfs );
  //  bfs.reset();

  cout << "Doing BFS from vertex 0 with a partition ... ";
  bfs.setRoot(0);
  int partitionMap[] = { 1, 1, 1, 3, 3, 3, 2, 2, 2};
  if (! bfs.enablePartitionRestrictions(partitionMap)) { 
    cout << "problem setting partition vector" << endl;
  } else if ( bfs.execute() ) {
    cout << " done. " << endl;
  } else {
    cout << " problem detected ! " << endl;
    cout << bfs.queryErrMsg() << endl;
  }
  printAll( bfs );
  bfs.disablePartitionRestrictions();
  //  bfs.reset();

  cout << "Doing BFS from vertex 0 with a partition and loose restrictions... ";
  bfs.setRoot(0);
  if (! bfs.enablePartitionRestrictions(partitionMap,BreadthFirstSearch::loose)) { 
    cout << "problem setting partition vector" << endl;
  } else if ( bfs.execute() ) {
    cout << " done. " << endl;
  } else {
    cout << " problem detected ! " << endl;
    cout << bfs.queryErrMsg() << endl;
  }
  printAll( bfs );
  bfs.disablePartitionRestrictions();
  //  bfs.reset();

  int partitionMap2[] = { 1, 1, 1, 3, 3, 3, 2, 2, 1};
  cout << "Doing BFS from vertex 0 with a new partition and loose restrictions... ";
  bfs.setRoot(0);
  if (! bfs.enablePartitionRestrictions(partitionMap2,BreadthFirstSearch::loose)) { 
    cout << "problem setting partition vector" << endl;
  } else if ( bfs.execute() ) {
    cout << " done. " << endl;
  } else {
    cout << " problem detected ! " << endl;
    cout << bfs.queryErrMsg() << endl;
  }
  printAll( bfs );
  bfs.disablePartitionRestrictions();
  //  bfs.reset();

  cout << "Doing BFS from vertex 0 with shortCircuit = 3 ... ";
  bfs.setRoot(0);
  if (! bfs.enableShortCircuiting(3) ) {
    cout << "problem setting shortCircuit = 3" << endl;
    cout << bfs.queryErrMsg() << endl;
  } else if ( bfs.execute() ) {
    if ( bfs.hasShortCircuited() ) {
      cout << "short-circuited!" << endl;
    } else {
      cout << " done. " << endl;
    }
  } else {
    cout << " problem detected ! " << endl;
    cout << bfs.queryErrMsg() << endl;
  }
  printAll( bfs );
  //  bfs.reset();
  
  cout << "Doing BFS from vertex 0 with shortCircuit = 2 ... ";
  bfs.setRoot(0);
  if (! bfs.enableShortCircuiting(2) ) {
    cout << "problem setting shortCircuit = 2" << endl;
    cout << bfs.queryErrMsg() << endl;
  } else if ( bfs.execute() ) {
    if ( bfs.hasShortCircuited() ) {
      cout << "short-circuited!" << endl;
    } else {
      cout << " done. " << endl;
    }
  } else {
    cout << " problem detected ! " << endl;
    cout << bfs.queryErrMsg() << endl;
  }
  printAll( bfs );
}

#define __FUNC__ "static void printAll(const BreadthFirstSearch& bfs)"
static void printAll(const BreadthFirstSearch& bfs) {
  FENTER;
  int nVisited = bfs.queryNVisited();
  static stopwatch timer;

  cout << "  BFS.nVisited    = " << bfs.queryNVisited() << endl;
  cout << "  BFS.nFarthest   = " << bfs.queryNFarthest() << endl;
  cout << "  BFS.height      = " << bfs.queryHeight() << endl;
  cout << "  BFS.width       = " << bfs.queryWidth() << endl;
//  cout << "  BFS.userTime    = " << (float) bfs.getTimer().queryTotalUserTicks()/timer.queryTicksPerSec() << endl;
//  cout << "  BFS.systemTime  = " << (float) bfs.getTimer().queryTotalSystemTicks()/timer.queryTicksPerSec() << endl;
  cout << "  BFS.maxDegree   = " << bfs.queryMaxDegree() << endl;
  cout << "  BFS.maxDegreeVtx= " << bfs.queryMaxDegreeVtx() << endl;
  cout << "  BFS.minDegree   = " << bfs.queryMinDegree() << endl;
  cout << "  BFS.minDegreeVtx= " << bfs.queryMinDegreeVtx() << endl;

  FCALL const SharedArray<int>& perm = bfs.getNew2Old();
  cout << "  BFS.new2old = " << endl;
  { for(int i=0; i<nVisited; i++) {
    cout <<"            [ " << i << "] = " << perm[i] << ",  dist = " << bfs.queryDistance(perm[i]) << endl;
  }}
  cout << endl;
  FEXIT;
}
#undef __FUNC__
