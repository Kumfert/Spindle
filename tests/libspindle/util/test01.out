Testing the adjacency iterators over the graph
adj(0) :1, 3, 
adj(1) :0, 2, 4, 
adj(2) :1, 5, 
adj(3) :0, 4, 6, 
adj(4) :1, 3, 5, 7, 
adj(5) :2, 4, 8, 
adj(6) :3, 7, 
adj(7) :4, 6, 8, 
adj(8) :5, 7, 

Creating the Breadth First Search (BFS) object ...  done. 
  BFS.nVisited    = 0
  BFS.nFarthest   = 0
  BFS.height      = 0
  BFS.width       = 0
  BFS.maxDegree   = -1
  BFS.maxDegreeVtx= -1
  BFS.minDegree   = -1
  BFS.minDegreeVtx= -1
  BFS.new2old = 

Doing BFS from vertex 2 ...  done. 
  BFS.nVisited    = 9
  BFS.nFarthest   = 1
  BFS.height      = 4
  BFS.width       = 3
  BFS.maxDegree   = 4
  BFS.maxDegreeVtx= 4
  BFS.minDegree   = 2
  BFS.minDegreeVtx= 2
  BFS.new2old = 
            [ 0] = 2,  dist = 0
            [ 1] = 1,  dist = 1
            [ 2] = 5,  dist = 1
            [ 3] = 0,  dist = 2
            [ 4] = 4,  dist = 2
            [ 5] = 8,  dist = 2
            [ 6] = 3,  dist = 3
            [ 7] = 7,  dist = 3
            [ 8] = 6,  dist = 4

Doing BFS from vertex 0 with a partition ...  done. 
  BFS.nVisited    = 3
  BFS.nFarthest   = 1
  BFS.height      = 2
  BFS.width       = 1
  BFS.maxDegree   = 3
  BFS.maxDegreeVtx= 1
  BFS.minDegree   = 2
  BFS.minDegreeVtx= 0
  BFS.new2old = 
            [ 0] = 0,  dist = 0
            [ 1] = 1,  dist = 1
            [ 2] = 2,  dist = 2

Doing BFS from vertex 0 with a partition and loose restrictions...  done. 
  BFS.nVisited    = 6
  BFS.nFarthest   = 1
  BFS.height      = 3
  BFS.width       = 2
  BFS.maxDegree   = 4
  BFS.maxDegreeVtx= 4
  BFS.minDegree   = 2
  BFS.minDegreeVtx= 0
  BFS.new2old = 
            [ 0] = 0,  dist = 0
            [ 1] = 1,  dist = 1
            [ 2] = 3,  dist = 1
            [ 3] = 2,  dist = 2
            [ 4] = 4,  dist = 2
            [ 5] = 5,  dist = 3

Doing BFS from vertex 0 with a new partition and loose restrictions...  done. 
  BFS.nVisited    = 8
  BFS.nFarthest   = 1
  BFS.height      = 5
  BFS.width       = 2
  BFS.maxDegree   = 4
  BFS.maxDegreeVtx= 4
  BFS.minDegree   = 2
  BFS.minDegreeVtx= 0
  BFS.new2old = 
            [ 0] = 0,  dist = 0
            [ 1] = 1,  dist = 1
            [ 2] = 3,  dist = 1
            [ 3] = 2,  dist = 2
            [ 4] = 4,  dist = 2
            [ 5] = 5,  dist = 3
            [ 6] = 8,  dist = 4
            [ 7] = 7,  dist = 5

Doing BFS from vertex 0 with shortCircuit = 3 ...  done. 
  BFS.nVisited    = 9
  BFS.nFarthest   = 1
  BFS.height      = 4
  BFS.width       = 3
  BFS.maxDegree   = 4
  BFS.maxDegreeVtx= 4
  BFS.minDegree   = 2
  BFS.minDegreeVtx= 0
  BFS.new2old = 
            [ 0] = 0,  dist = 0
            [ 1] = 1,  dist = 1
            [ 2] = 3,  dist = 1
            [ 3] = 2,  dist = 2
            [ 4] = 4,  dist = 2
            [ 5] = 6,  dist = 2
            [ 6] = 5,  dist = 3
            [ 7] = 7,  dist = 3
            [ 8] = 8,  dist = 4

Doing BFS from vertex 0 with shortCircuit = 2 ... short-circuited!
  BFS.nVisited    = 6
  BFS.nFarthest   = 3
  BFS.height      = 2
  BFS.width       = 3
  BFS.maxDegree   = 3
  BFS.maxDegreeVtx= 1
  BFS.minDegree   = 2
  BFS.minDegreeVtx= 0
  BFS.new2old = 
            [ 0] = 0,  dist = 0
            [ 1] = 1,  dist = 1
            [ 2] = 3,  dist = 1
            [ 3] = 2,  dist = 2
            [ 4] = 4,  dist = 2
            [ 5] = 6,  dist = 2

